### PR TITLE
[TwoAddressInstruction] Recompute live intervals for partial defs

### PR DESCRIPTION
**Pull request by perlfu**
Force live interval recomputation for a register if its definition is narrowed to become partial.  The live interval repair process cannot otherwise detect these changes.

### CODE REVIEW

File: llvm/lib/CodeGen/TwoAddressInstructionPass.cpp [hunk 1]
```diff
@@ -1937,13 +1937,16 @@ eliminateRegSequence(MachineBasicBlock::iterator &MBBI) {
...
   bool DefEmitted = false;
+  bool DefIsPartial = false;
   for (unsigned i = 1, e = MI.getNumOperands(); i < e; i += 2) {
     MachineOperand &UseMO = MI.getOperand(i);
     Register SrcReg = UseMO.getReg();
     unsigned SubIdx = MI.getOperand(i+1).getImm();
     // Nothing needs to be inserted for undef operands.
-    if (UseMO.isUndef())
+    if (UseMO.isUndef()) {
+      DefIsPartial = true;
```
Comment (arsenm) [inline llvm/lib/CodeGen/TwoAddressInstructionPass.cpp:1947]: Do we actually know a REG_SEQUENCE fully covers the result register? e.g. something like 

`%2:vreg_128 = REG_SEQUENCE %0, %subreg.sub0, %1, %subreg.sub1`

is valid but doesn't define all lanes

**Comment by jayfoad** [comment]
An idea for testing this: you could add an `-early-live-intervals` RUN line to some or all of the existing codegen tests that would be fixed by this patch. I think they are:
```
  LLVM :: CodeGen/AMDGPU/GlobalISel/add.vni16.ll
  LLVM :: CodeGen/AMDGPU/ds_gws_align.ll
  LLVM :: CodeGen/AMDGPU/llvm.amdgcn.image.atomic.dim.ll
  LLVM :: CodeGen/AMDGPU/llvm.amdgcn.image.dim.gfx90a.ll
  LLVM :: CodeGen/AMDGPU/llvm.amdgcn.image.sample.dim.gfx90a.ll
```

File: llvm/lib/CodeGen/TwoAddressInstructionPass.cpp (hunk 1 above)
Comment (perlfu) [reply llvm/lib/CodeGen/TwoAddressInstructionPass.cpp:1947]: Yes, those should have already been handled.

File: llvm/test/CodeGen/AMDGPU/early-lis-two-address-partial-def.mir [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 4
+# RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx90a -run-pass=liveintervals -run-pass=twoaddressinstruction -verify-machineinstrs -o - %s | FileCheck --check-prefix=GFX90A %s
```
[the remark this answers is not in the document]
Comment (perlfu) [reply llvm/test/CodeGen/AMDGPU/early-lis-two-address-partial-def.mir:2]: It is required as the verifier detects the live interval errors that occur in this test.

**Review by arsenm — APPROVED** [review]

**Pull request merged**